_posts/2025-02-10-bash_scripting_an_overview.md

```diff
@@ -0,0 +1,163 @@
+
+To encapsulate the logging within a separate script, you can use a wrapper script like this. Here is the `logger.sh` script:
+
+```bash
+#!/bin/bash
+# ... (comments) ...
+# Options:
+#   $1: Path to the script to be logged.
+
+"$1" > "${1}_output.txt" 2> "${1}_errors.txt"
+```
+<a href="https://www.youtube.com/watch?v=1pQ527fGhVQ&t=1298.28s" target="_blank"><img src="/images/blogs/rebecca_linux_image_at_1298.28.png" alt=" OK, so I've already prepared an example of logging in a script called logger." width="450"/></a>
+
+This `logger.sh` script takes the path to another script (passed as the first positional parameter, `$1`) and executes it, redirecting output and errors to files with appropriate names.
+
+## Enhancing the Script: Adding Options
+
+Using positional parameters (`$1`, `$2`, etc.) can be fragile because the order matters. Options (e.g., `-f`, `-d`) provide a more robust and user-friendly way to configure script behavior. The `getopts` built-in command helps parse options.
+
+Here's a modified version of the script that accepts an `-f` option to specify the columns to extract:
+
+```bash
+#!/bin/bash
+# ... (previous comments) ...
+# Options:
+#   -f <fields>: Comma-separated list of fields to extract.
+
+while getopts "f:" opt; do
+  case $opt in
+    f)
+      fields="$OPTARG"
+      ;;
+    \?)
+      echo "Usage: $0 [-f <fields>]"
+      exit 1
+      ;;
+  esac
+done
+
+URL="http://www.econ.yale.edu/~shiller/data/ie_data.xls"
+echo "Downloading data from $URL"
+curl -O "$URL"
+ssconvert -S stock_data.xlsx stock_data.csv
+cut -d, -f"$fields" stock_data.csv.0 | head
+rm stock_data*
+```
+
+<a href="https://www.youtube.com/watch?v=1pQ527fGhVQ&t=243.1s" target="_blank"><img src="/images/blogs/rebecca_linux_image_at_243.10.png" alt=" OK, so I've already prepared a script process with options that has options in them." width="450"/></a>
+
+This version uses `getopts "f:" opt` to parse the `-f` option. [ In line 16, that string f colon tells you about the valid options.](https://www.youtube.com/watch?v=1pQ527fGhVQ&t=537s) The colon after `f` indicates that it requires an argument.  The `case` statement handles the option: if `-f` is provided, its value (accessed via `$OPTARG`) is stored in the `fields` variable. [ So fields would be 10 or 10,1.](https://www.youtube.com/watch?v=1pQ527fGhVQ&t=564s) If an invalid option is provided, a usage message is printed, and the script exits. The `cut` command now uses `-f"$fields"` to extract the specified columns. The `esac` closes out the case statement. [ When you create a lot of these control structures, like you type case to start it, and then ESAC, which is case backwards, to end the case statement, to tell Bash, you know, case part is done.](https://www.youtube.com/watch?v=1pQ527fGhVQ&t=982s)
+
+These examples show how to build upon a simple Bash script to add functionality, improve robustness, and make it more user-friendly.
```

Comment:
Can we add resources at the end? There are links to the resources in the video description on youtube.
Resources. The resources would be links to the repo, slides and the bash file examples.